client.go

```diff
@@ -1577,17 +1589,65 @@ func dialAddr(addr string, dial DialFunc, dialDualStack, isTLS bool, tlsConfig *
 		}
 		addr = addMissingPort(addr, isTLS)
 	}
-	conn, err := dial(addr)
+	rawConn, err := dial(addr)
 	if err != nil {
 		return nil, err
 	}
-	if conn == nil {
+	if rawConn == nil {
 		panic("BUG: DialFunc returned (nil, nil)")
 	}
 	if isTLS {
-		conn = tls.Client(conn, tlsConfig)
+		var (
+			errChannel chan error
+			timer      *time.Timer
+		)
+
+		if timeout != 0 {
+			errChannel = make(chan error, 2)
+			timer = time.AfterFunc(timeout, func() {
+				errChannel <- timeoutError{}
+			})
```

Comment:
Same here, you can use `AcquireTimer()`.

Reply:
Done

Reply:
time.AfterFunc runs goroutine only in case when timer fired, with AcquireTimer we need to create additional goroutine anyway